File: examplesForThesis-R/ch4-example-gp-priorrealisations.R
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GP kernels
# Example 4.5 in Section 4.4.1
#
########################################################################

dx = read.table("ch4-example-gp-priorrealisations-x.csv",sep=",",header=T)[,-1]
dz1 = read.table("ch4-example-gp-priorrealisations-z1.csv",sep=",",header=T)[,-1]
dz2 = read.table("ch4-example-gp-priorrealisations-z2.csv",sep=",",header=T)[,-1]
dz3 = read.table("ch4-example-gp-priorrealisations-z3.csv",sep=",",header=T)[,-1]
dz4 = read.table("ch4-example-gp-priorrealisations-z4.csv",sep=",",header=T)[,-1]
dz5 = read.table("ch4-example-gp-priorrealisations-z5.csv",sep=",",header=T)[,-1]
dz6 = read.table("ch4-example-gp-priorrealisations-z6.csv",sep=",",header=T)[,-1]

layout(matrix(1:6, 3, 2, byrow = FALSE)); par(mar=c(4,5,1,1))

plot(dx,dz1[,1],type="l",lwd=3,xlab="x",ylab="f(x)",ylim=c(-3,3),col="darkgreen"); lines(dx,dz1[,2],lwd=3,col="darkgreen"); lines(dx,dz1[,3],lwd=3,col="darkgreen"); text(10,3,"SE, l=1",pos=2);
plot(dx,dz3[,1],type="l",lwd=3,xlab="x",ylab="f(x)",ylim=c(-3,3),col="darkred"); lines(dx,dz3[,2],lwd=3,col="darkred"); lines(dx,dz3[,3],lwd=3,col="darkred"); text(10,3,"Matérn 5/2, l=1",pos=2);
plot(dx,dz2[,1],type="l",lwd=3,xlab="x",ylab="f(x)",ylim=c(-3,3),col="steelblue"); lines(dx,dz2[,2],lwd=3,col="steelblue"); lines(dx,dz2[,3],lwd=3,col="steelblue"); text(10,3,"Matérn 3/2, l=1",pos=2);
plot(dx,dz4[,1],type="l",lwd=3,xlab="x",ylab="f(x)",ylim=c(-3,3),col="darkgreen"); lines(dx,dz4[,2],lwd=3,col="darkgreen"); lines(dx,dz4[,3],lwd=3,col="darkgreen"); text(10,3,"SE, l=3",pos=2);
plot(dx,dz6[,1],type="l",lwd=3,xlab="x",ylab="f(x)",ylim=c(-3,3),col="darkred"); lines(dx,dz6[,5],lwd=3,col="darkred"); lines(dx,dz6[,6],lwd=3,col="darkred"); text(10,3,"Matérn 5/2, l=3",pos=2);
plot(dx,dz5[,1],type="l",lwd=3,xlab="x",ylab="f(x)",ylim=c(-3,3),col="steelblue"); lines(dx,dz5[,2],lwd=3,col="steelblue"); lines(dx,dz5[,3],lwd=3,col="steelblue"); text(10,3,"Matérn 3/2, l=3",pos=2);

########################################################################
# End of file
########################################################################

File: examplesForThesis-python/ch4-example-gp-priorrealisations.py
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GP kernels
# Example 4.5 in Section 4.4.1
#
########################################################################

import numpy as np
import pylab as pb
import GPy
import pandas

# Generate a grid to evaluate the kernels on
d  = 1;
X  = np.linspace(0.,10.,500); X = X[:,None];
mu = np.zeros((500)); 

# Use length scale 1 and generate a realisation
k = GPy.kern.rbf(d, lengthscale=1);      C = k.K(X,X); Z1 = np.random.multivariate_normal(mu,C,10)
k = GPy.kern.Matern32(d, lengthscale=1); C = k.K(X,X); Z2 = np.random.multivariate_normal(mu,C,10)
k = GPy.kern.Matern52(d, lengthscale=1); C = k.K(X,X); Z3 = np.random.multivariate_normal(mu,C,10)

# Use length scale 2 and generate a realisation
k = GPy.kern.rbf(d, lengthscale=2);      C = k.K(X,X); Z4 = np.random.multivariate_normal(mu,C,10)
k = GPy.kern.Matern32(d, lengthscale=2); C = k.K(X,X); Z5 = np.random.multivariate_normal(mu,C,10)
k = GPy.kern.Matern52(d, lengthscale=2); C = k.K(X,X); Z6 = np.random.multivariate_normal(mu,C,10)

# Export data to R for plotting
pandas.DataFrame(X).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-x.csv");
pandas.DataFrame(Z1.transpose()).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-z1.csv");
pandas.DataFrame(Z2.transpose()).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-z2.csv");
pandas.DataFrame(Z3.transpose()).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-z3.csv");
pandas.DataFrame(Z4.transpose()).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-z4.csv");
pandas.DataFrame(Z5.transpose()).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-z5.csv");
pandas.DataFrame(Z6.transpose()).to_csv("../figures/ch4-example-gp-priorrealisations/ch4-example-gp-priorrealisations-z6.csv");

########################################################################
# End of file
########################################################################